packages/babel-compiler/babel-compiler.js

```diff
@@ -216,40 +222,58 @@ BCp._inferHelper = function (inputFile, babelOptions, babelrc) {
     function req(id) {
       var isTopLevel = "./".indexOf(id.charAt(0)) < 0;
       var presetOrPlugin;
+      var presetOrPluginMeta;
 
       if (isTopLevel) {
         try {
           // If the identifier is top-level, try to prefix it with
           // "babel-plugin-" or "babel-preset-".
           presetOrPlugin = inputFile.require(prefix + id);
+          presetOrPluginMeta = inputFile.require(
+            packageNameFromTopLevelModuleId(prefix + id) + '/package.json');
         } catch (e) {
           if (e.code !== "MODULE_NOT_FOUND") {
             throw e;
           }
           // Fall back to requiring the plugin as-is if the prefix failed.
           presetOrPlugin = inputFile.require(id);
+          presetOrPluginMeta = inputFile.require(
+            packageNameFromTopLevelModuleId(id) + '/package.json');
         }
       } else {
         // If the identifier is not top-level, but relative or absolute,
         // then it will be required as-is, so that you can implement your
         // own Babel plugins locally, rather than always using plugins
         // installed from npm.
         presetOrPlugin = inputFile.require(id);
```

Comment:
There's also `inputFile.resolve(id)`, which could be used to determine the full absolute path of the module (with suitable conversion for Windows), and then we could hash the file contents instead of hashing the stringified function.